### PR TITLE
ENH: add methods to walk AD plugins

### PR DESCRIPTION
**Pull request by tacaswell**
- find all plugins that ophyd knows about
 - find a plugin by port name
 - validate that the full processing path is know to ophyd
 - turn blocking_callbacks on by default in all Plugin stage methods

### CODE REVIEW

**Comment by klauer** [comment]
Need to add networkx to dependency list

File: ophyd/areadetector/base.py [hunk 1]
```diff
@@ -148,3 +149,110 @@ def default_match(attr, signal, doc):
...
+        signames = self._sub_devices
+
+        for nm in signames:
+            sig = getattr(self, nm)
+            if hasattr(sig, 'get_port_dictionary'):
+                ret.update(sig.get_port_dictionary())
+
+        return ret
+
+    def get_asyn_digraph(self):
```
Comment (klauer) [inline ophyd/areadetector/base.py]: (`port` or `asyn`) -> `asyn_port` everywhere, or these method names won't make much sense

File: ophyd/areadetector/plugins.py [hunk 2]
```diff
@@ -106,6 +112,15 @@ def array_pixels(self):
 
         return pixels
 
+    def read_configuration(self):
+        ret = super().read_configuration()
+
+        source_port = self.nd_array_port.get()
+        source_plugin = self.root.get_plugin_by_port(source_port)
+        ret[source_port] = source_plugin.read_configuration()
```
Comment (klauer) [inline ophyd/areadetector/plugins.py]: I like this idea, but won't this turn into:
{'port': dict(attr=value), ...}
where we actually want just a single merged dictionary?

Comment (tacaswell) [inline ophyd/areadetector/plugins.py]: The idea was to to directly encode the ordering in the nesting.  Get the configuration of the parent via `dd[dd['nd_array_port']]`.

I do not think we want to merge these dictionaries as I think eventually will have `understand_what_stats_plugin_was_doing()` functions on the analysis side and it would be nice to not have to allow those to accept (but ignore) extra keys.

File: ophyd/areadetector/base.py (hunk 1 above)
Comment (tacaswell) [inline ophyd/areadetector/base.py]: fair.  I am bad at naming things.

File: ophyd/areadetector/plugins.py (hunk 2 above)
Comment (klauer) [inline ophyd/areadetector/plugins.py]: ... but this breaks the `read_configuration` API? Call it something else?

Comment (tacaswell) [inline ophyd/areadetector/plugins.py]: how does this break the API?

Comment (tacaswell) [inline ophyd/areadetector/plugins.py]: This breaks api because we do not have a way of doing nested descriptors.

File: ophyd/quadem.py [hunk 3]
```diff
@@ -16,7 +16,9 @@ def _current_fields(attr_base, field_base, range_, **kwargs):
 
     return defn
 
+
 class QuadEM(SingleTrigger, DetectorBase):
+    port_name = Cpt(Signal, value='NSLS2_EM')
```
Comment (klauer) [inline ophyd/quadem.py]: Mind leaving a comment here as to why this is necessary?

It's also a bit concerning that the QuadEM IOC is re-used for other devices than the NSLS2-EM. I don't know what their port names are, nor if they actually have an EPICS PV in their databases with the port name exposed...

**Comment by klauer** [comment]
Automatically making mbbi/mbbo and bi/bo records (the ones you would consider to be enums) use `string=True` is something I wish I had done with the class generation from template. Thanks for tweaking some of them.

File: tests/test_areadetector.py [hunk 4]
```diff
@@ -116,6 +117,64 @@ class MyDetector(SimDetector):
...
+
+    assert det.tiff1 is det.get_plugin_by_asyn_port(det.tiff1.port_name.get())
+    assert det.cam is det.get_plugin_by_asyn_port(det.cam.port_name.get())
+    assert det.roi1 is det.get_plugin_by_asyn_port(det.roi1.port_name.get())
+
+
+def test_read_configuration_smoke():
+    class MyDetector(SingleTrigger, SimDetector):
+        stats1 = Cpt(StatsPlugin, 'Stats1:')
+        porc1 = Cpt(ProcessPlugin, 'ROI1:')
```
Comment (klauer) [inline tests/test_areadetector.py]: Typo here causing test to fail

File: ophyd/areadetector/plugins.py [hunk 5]
```diff
@@ -106,6 +112,24 @@ def array_pixels(self):
 
         return pixels
 
+    def read_configuration(self):
+        ret = super().read_configuration()
+
+        source_port = self.nd_array_port.get()
+        source_plugin = self.root.get_plugin_by_asyn_port(source_port)
```
Comment (klauer) [inline ophyd/areadetector/plugins.py]: `source_plugin` seems like something that'd be handy to have as a property (or a `get_*` as you seem to prefer)

File: ophyd/areadetector/base.py [hunk 6]
```diff
@@ -148,3 +149,110 @@ def default_match(attr, signal, doc):
...
+                    return sig
+        return None
+
+    def get_asyn_port_dictionary(self):
+        '''Return port name : component map
+
+        Returns
+        -------
+        port_map : dict
+            keyed on port name (str) values are OphydObjects
```
Comment (klauer) [inline ophyd/areadetector/base.py]: Yes, they are `OphydObj`ects, but I think we can be a bit more specific here

File: tests/test_areadetector.py (hunk 4 above)
Comment (tacaswell) [inline tests/test_areadetector.py]: > (the ones you would consider to be enums)

I found them by looking for descriptions that and an `enum_str` field in them.  Is calling them enums not fair?

**Comment by klauer** [comment]
Overall 👍 Great addition to the current functionality

File: tests/test_areadetector.py (hunk 4 above)
Comment (klauer) [inline tests/test_areadetector.py]: It's fair given the view ophyd gives - but they're represented by different records at the EPICS level. So from that point of view, calling them all just enums is strange.

**Comment by danielballan** [comment]
+1 for our first usage of networkx

**Comment by tacaswell** [comment]
@klauer The second half of this is ready to review.

File: ophyd/areadetector/base.py [hunk 7]
```diff
@@ -148,3 +150,117 @@ def default_match(attr, signal, doc):
...
+
+        Returns
+        -------
+        ret : ADBase or None
+            Either the requested plugin or None if not found
+
+        '''
+        try:
+            name = self.port_name.get()
+            if name == port_name:
```
Comment (klauer) [inline ophyd/areadetector/base.py]: Bike-shedding: move this out of the try/except block, `signames` variable could be easily removed here and below (unless you get a bonus for line counts then by all means leave it in)

File: ophyd/areadetector/base.py [hunk 8]
```diff
@@ -148,3 +150,117 @@ def default_match(attr, signal, doc):
...
+
+        port_map : dict
+            Mapping between port_name and ADBase objects
+        '''
+        port_map = self.get_asyn_port_dictionary()
+        G = nx.DiGraph()
+        for out_port, cpt in port_map.items():
+            try:
+                in_port = cpt.nd_array_port.get()
+                G.add_edge(in_port, out_port)
```
Comment (klauer) [inline ophyd/areadetector/base.py]: `G.add_edge(...)` should go in an `else` clause to be clear what the `except AttributeError` is trying to catch

File: ophyd/areadetector/base.py [hunk 9]
```diff
@@ -148,3 +150,117 @@ def default_match(attr, signal, doc):
...
+    def missing_plugins(self):
+        g, port_map = self.get_asyn_digraph()
+        ret = []
+
+        for node in g:
+            if node not in port_map:
+                ret.append(node)
+
+        return ret
+    config_names = Component(AttributeSignal, '_configuration_names',
```
Comment (klauer) [inline ophyd/areadetector/base.py]: `configuration_attrs` but `config_names`... probably should keep it consistent?

Comment (tacaswell) [inline ophyd/areadetector/base.py]: so consistent long it is

Comment (klauer) [inline ophyd/areadetector/base.py]: `add_prefix` is also unnecessary, assuming you use `attr='_configuration_names'` instead of the positional argument here

File: ophyd/areadetector/plugins.py [hunk 10]
```diff
@@ -59,14 +59,30 @@ def register_plugin(cls):
...
+        if configuration_attrs is None:
+            configuration_attrs = self._default_configuration_attrs
         # Turn array callbacks on during staging.
         # Without this, no array data is sent to the plugins.
-        super().__init__(*args, **kwargs)
-
+        super().__init__(*args, configuration_attrs=configuration_attrs,
+                         **kwargs)
+        # make sure it is the right type of plugin
+        if not self.plugin_type.get().startswith(self._plugin_type):
```
Comment (klauer) [inline ophyd/areadetector/plugins.py]: It may be useful to remove this check in some cases. I should be able to use `PluginBase` with any plugin and not always the exact plugin class.

To that end, can you add a `if self._plugin_type is not None and ...` before this?

File: ophyd/areadetector/plugins.py [hunk 11]
```diff
@@ -106,6 +122,43 @@ def array_pixels(self):
...
 
+    def read_configuration(self):
+        ret = super().read_configuration()
+
+        ret.update(self.source_plugin.read_configuration())
+
+        return ret
+
+    @property
+    def source_plugin(self):
```
Comment (klauer) [inline ophyd/areadetector/plugins.py]: A property - great idea! 👍 But please include docstrings...

File: ophyd/areadetector/plugins.py [hunk 12]
```diff
@@ -106,6 +122,43 @@ def array_pixels(self):
...
+        source_port = self.nd_array_port.get()
+        source_plugin = self.root.get_plugin_by_asyn_port(source_port)
+        return source_plugin
+
+    def describe_configuration(self):
+        ret = super().describe_configuration()
+
+        source_port = self.nd_array_port.get()
+        source_plugin = self.root.get_plugin_by_asyn_port(source_port)
```
Comment (klauer) [inline ophyd/areadetector/plugins.py]: Use the source_plugin property here?

File: ophyd/areadetector/base.py (hunk 7 above)
Comment (tacaswell) [inline ophyd/areadetector/base.py]: I changed this, but I do not think in the way you meant.

File: ophyd/utils/epics_pvs.py [hunk 13]
```diff
@@ -428,11 +428,12 @@ def _compare_maybe_enum(a, b, enums, atol, rtol):
 
 
 _type_map = {'number': (float, ),
-             'array': (np.ndarray, ),
+             'array': (np.ndarray, list),
```
Comment (klauer) [inline ophyd/utils/epics_pvs.py]: I guess this is fine, but I have a feeling it was working around what `ArrayAttributeSignal` was made for: https://github.com/NSLS-II/ophyd/blob/master/ophyd/signal.py#L941

Comment (tacaswell) [inline ophyd/utils/epics_pvs.py]: and I actually bet that on return this is going to be turned into an object array as we upcast everything that is 'dtype' array to `np.array`.

I should either revert these changes and use `ArrayAttributeSignal` or we make list a first class thing in the schema

File: ophyd/areadetector/base.py (hunk 7 above)
Comment (klauer) [inline ophyd/areadetector/base.py]:
```python
        try:
            name = self.port_name.get()
        except AttributeError:
            pass

        if name == port_name:
            return self

        for nm in self._sub_devices:
            cpt = getattr(self, nm)
            if hasattr(cpt, 'get_plugin_by_asyn_port'):
                sig = cpt.get_plugin_by_asyn_port(port_name)
                if sig is not None:
                    return sig
        return None
```

**Comment by klauer** [comment]
Passes travis finally 👍

**Comment by klauer** [comment]
My last remaining quibble: `_default_configuration_attrs` is not being checked for all plugins. There should be a test that ensures all of the default attributes exist.

**Comment by tacaswell** [comment]
Why did we lose codecov in this?

**Comment by klauer** [comment]
👍
